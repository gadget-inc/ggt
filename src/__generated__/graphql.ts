/**
 * ======================================================
 * THIS IS A GENERATED FILE! DO NOT EDIT IT MANUALLY!
 *
 * You can regenerate it by running `npm run generate-graphql`.
 * ======================================================
 */

export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string | number; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
  /** A date string, such as 2007-12-03, compliant with the `full-date` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar. */
  Date: { input: any; output: any; }
  /** A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the `date-time` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar. */
  DateTime: { input: any; output: any; }
  JSON: { input: { [key: string]: any }; output: { [key: string]: any }; }
  /** The `Upload` scalar type represents a file upload. */
  Upload: { input: any; output: any; }
};

export type ApiUpgradeConvergePlanResult = {
  __typename?: 'APIUpgradeConvergePlanResult';
  items: Array<Scalars['JSON']['output']>;
  success: Scalars['Boolean']['output'];
};

export type ChangeAppDomainResult = {
  __typename?: 'ChangeAppDomainResult';
  onlyValidate?: Maybe<Scalars['Boolean']['output']>;
  reason?: Maybe<Scalars['String']['output']>;
  success: Scalars['Boolean']['output'];
};

export type ContributorResult = {
  __typename?: 'ContributorResult';
  email: Scalars['String']['output'];
  isOwner: Scalars['Boolean']['output'];
  isPending: Scalars['Boolean']['output'];
};

export type DeleteAppStatusResult = {
  __typename?: 'DeleteAppStatusResult';
  isNotCreator?: Maybe<Scalars['Boolean']['output']>;
  isNotOwner?: Maybe<Scalars['Boolean']['output']>;
  success: Scalars['Boolean']['output'];
};

export type EnableFrontendResult = {
  __typename?: 'EnableFrontendResult';
  reason?: Maybe<Scalars['String']['output']>;
  success: Scalars['Boolean']['output'];
};

export type EnvironmentPatchResult = {
  __typename?: 'EnvironmentPatchResult';
  success: Scalars['Boolean']['output'];
};

export type EnvironmentPublishResult = {
  __typename?: 'EnvironmentPublishResult';
  success: Scalars['Boolean']['output'];
};

export type EnvironmentSubscriptionResult = {
  __typename?: 'EnvironmentSubscriptionResult';
  patches: Array<Scalars['JSON']['output']>;
};

export type EnvironmentTreeClientId = {
  clientType: Scalars['String']['input'];
  id: Scalars['String']['input'];
};

export type FileSyncChangedEvent = {
  __typename?: 'FileSyncChangedEvent';
  content: Scalars['String']['output'];
  encoding: FileSyncEncoding;
  mode: Scalars['Float']['output'];
  path: Scalars['String']['output'];
};

export type FileSyncChangedEventInput = {
  content: Scalars['String']['input'];
  encoding?: InputMaybe<FileSyncEncoding>;
  mode: Scalars['Float']['input'];
  oldPath?: InputMaybe<Scalars['String']['input']>;
  path: Scalars['String']['input'];
};

export type FileSyncDeletedEvent = {
  __typename?: 'FileSyncDeletedEvent';
  path: Scalars['String']['output'];
};

export type FileSyncDeletedEventInput = {
  path: Scalars['String']['input'];
};

export enum FileSyncEncoding {
  Base64 = 'base64',
  Utf8 = 'utf8'
}

export type GadgetRole = {
  __typename?: 'GadgetRole';
  key: Scalars['String']['output'];
  name: Scalars['String']['output'];
  order: Scalars['Int']['output'];
  selectable: Scalars['Boolean']['output'];
};

export type IdentifySupportConversationResult = {
  __typename?: 'IdentifySupportConversationResult';
  identificationEmail: Scalars['String']['output'];
  identificationToken: Scalars['String']['output'];
};

export type LogSearchResult = {
  __typename?: 'LogSearchResult';
  data: Scalars['JSON']['output'];
  status: Scalars['String']['output'];
};

export type MigrateEnvironmentsResult = {
  __typename?: 'MigrateEnvironmentsResult';
  success: Scalars['Boolean']['output'];
};

export type Mutation = {
  __typename?: 'Mutation';
  changeAppDomain?: Maybe<ChangeAppDomainResult>;
  deleteApp?: Maybe<DeleteAppStatusResult>;
  enableFrontend?: Maybe<EnableFrontendResult>;
  migrateEnvironments?: Maybe<MigrateEnvironmentsResult>;
  patchEnvironmentTree?: Maybe<EnvironmentPatchResult>;
  publish?: Maybe<EnvironmentPublishResult>;
  publishFileSyncEvents: PublishFileSyncEventsResult;
  refreshScopes?: Maybe<RefreshScopesResult>;
  registerWebhooks?: Maybe<RegisterWebhooksResult>;
  removeContributor?: Maybe<RemoveContributorResult>;
  sendAppInvitation?: Maybe<SendAppInvitationResult>;
  unregisterWebhooks?: Maybe<UnregisterWebhooksResult>;
  uploadFiles: UploadFilesResult;
};


export type MutationChangeAppDomainArgs = {
  newSubdomain: Scalars['String']['input'];
  onlyValidate?: InputMaybe<Scalars['Boolean']['input']>;
};


export type MutationEnableFrontendArgs = {
  hasShopifyConnection: Scalars['Boolean']['input'];
};


export type MutationMigrateEnvironmentsArgs = {
  existingToProduction: Scalars['Boolean']['input'];
};


export type MutationPatchEnvironmentTreeArgs = {
  clientID: EnvironmentTreeClientId;
  patches: Array<Scalars['JSON']['input']>;
};


export type MutationPublishFileSyncEventsArgs = {
  input: PublishFileSyncEventsInput;
};


export type MutationRefreshScopesArgs = {
  appConfigKey: Scalars['String']['input'];
  connectionKey: Scalars['String']['input'];
  shopId: Scalars['String']['input'];
};


export type MutationRegisterWebhooksArgs = {
  connectionKey: Scalars['String']['input'];
  keepExtraTopics?: InputMaybe<Scalars['Boolean']['input']>;
  modelKeys?: InputMaybe<Array<Scalars['String']['input']>>;
  shopIds: Array<Scalars['String']['input']>;
};


export type MutationRemoveContributorArgs = {
  email: Scalars['String']['input'];
  isInvitation: Scalars['Boolean']['input'];
};


export type MutationSendAppInvitationArgs = {
  email?: InputMaybe<Scalars['String']['input']>;
  emails?: InputMaybe<Array<Scalars['String']['input']>>;
  resend?: InputMaybe<Scalars['Boolean']['input']>;
};


export type MutationUnregisterWebhooksArgs = {
  apiKeys?: InputMaybe<Array<Scalars['String']['input']>>;
  connectionKey: Scalars['String']['input'];
  modelKeys?: InputMaybe<Array<Scalars['String']['input']>>;
};


export type MutationUploadFilesArgs = {
  files: Array<UploadFile>;
};

export type PublishFileSyncEventsInput = {
  changed: Array<FileSyncChangedEventInput>;
  deleted: Array<FileSyncDeletedEventInput>;
  expectedRemoteFilesVersion: Scalars['String']['input'];
};

export type PublishFileSyncEventsResult = {
  __typename?: 'PublishFileSyncEventsResult';
  remoteFilesVersion: Scalars['String']['output'];
};

export type Query = {
  __typename?: 'Query';
  apiUpgradeConvergePlan?: Maybe<ApiUpgradeConvergePlanResult>;
  currentUser: User;
  environmentTreeChildKeys: Array<Scalars['String']['output']>;
  environmentTreePath?: Maybe<Scalars['JSON']['output']>;
  identifySupportConversation?: Maybe<IdentifySupportConversationResult>;
  listContributors: Array<ContributorResult>;
  logsSearch: LogSearchResult;
  remoteFilesVersion: Scalars['String']['output'];
  roles: Array<GadgetRole>;
  runTestSupportFunction?: Maybe<Scalars['JSON']['output']>;
  team: TeamResult;
  typesManifest: TypesManifest;
};


export type QueryApiUpgradeConvergePlanArgs = {
  currentVersion: Scalars['String']['input'];
  targetVersion: Scalars['String']['input'];
};


export type QueryEnvironmentTreeChildKeysArgs = {
  path: Scalars['String']['input'];
};


export type QueryEnvironmentTreePathArgs = {
  hydrateChildrenGlobs?: InputMaybe<Array<Scalars['String']['input']>>;
  path: Scalars['String']['input'];
};


export type QueryLogsSearchArgs = {
  direction?: InputMaybe<Scalars['String']['input']>;
  end?: InputMaybe<Scalars['DateTime']['input']>;
  limit?: InputMaybe<Scalars['Int']['input']>;
  query: Scalars['String']['input'];
  start?: InputMaybe<Scalars['DateTime']['input']>;
  step?: InputMaybe<Scalars['Int']['input']>;
};


export type QueryTypesManifestArgs = {
  dependenciesHash: Scalars['String']['input'];
};

export type RefreshScopesResult = {
  __typename?: 'RefreshScopesResult';
  success: Scalars['Boolean']['output'];
};

export type RegisterWebhooksResult = {
  __typename?: 'RegisterWebhooksResult';
  success: Scalars['Boolean']['output'];
};

export type RemoteFileSyncEvents = {
  __typename?: 'RemoteFileSyncEvents';
  changed: Array<FileSyncChangedEvent>;
  deleted: Array<FileSyncDeletedEvent>;
  remoteFilesVersion: Scalars['String']['output'];
};

export type RemoveContributorResult = {
  __typename?: 'RemoveContributorResult';
  reason?: Maybe<Scalars['String']['output']>;
  success: Scalars['Boolean']['output'];
};

export type SendAppInvitationResult = {
  __typename?: 'SendAppInvitationResult';
  reason?: Maybe<Scalars['String']['output']>;
  success: Scalars['Boolean']['output'];
};

export type Subscription = {
  __typename?: 'Subscription';
  editorActive?: Maybe<Scalars['Boolean']['output']>;
  environmentTreePathPatches?: Maybe<EnvironmentSubscriptionResult>;
  logsSearch: LogSearchResult;
  remoteFileSyncEvents: RemoteFileSyncEvents;
  typesManifestStream: TypesManifest;
};


export type SubscriptionEnvironmentTreePathPatchesArgs = {
  clientID: EnvironmentTreeClientId;
  path: Scalars['String']['input'];
};


export type SubscriptionLogsSearchArgs = {
  limit?: InputMaybe<Scalars['Int']['input']>;
  query: Scalars['String']['input'];
  start?: InputMaybe<Scalars['DateTime']['input']>;
};


export type SubscriptionRemoteFileSyncEventsArgs = {
  encoding?: InputMaybe<FileSyncEncoding>;
  localFilesVersion: Scalars['String']['input'];
};

export type TeamMember = {
  __typename?: 'TeamMember';
  contributesToApp: Scalars['Boolean']['output'];
  email: Scalars['String']['output'];
};

export type TeamResult = {
  __typename?: 'TeamResult';
  availableSeats?: Maybe<Scalars['Int']['output']>;
  costPerSeat?: Maybe<Scalars['String']['output']>;
  teamMembers: Array<TeamMember>;
};

export type TypeManifestEntry = {
  __typename?: 'TypeManifestEntry';
  declaration: Scalars['String']['output'];
  path: Scalars['String']['output'];
};

export type TypesManifest = {
  __typename?: 'TypesManifest';
  cdn: Array<Scalars['String']['output']>;
  dependenciesHash: Scalars['String']['output'];
  entries: Array<TypeManifestEntry>;
  environmentVersion: Scalars['Int']['output'];
};

export type UnregisterWebhooksResult = {
  __typename?: 'UnregisterWebhooksResult';
  success: Scalars['Boolean']['output'];
};

export type UploadFile = {
  file: Scalars['Upload']['input'];
  path: Scalars['String']['input'];
};

export type UploadFilesResult = {
  __typename?: 'UploadFilesResult';
  success: Scalars['Boolean']['output'];
};

export type User = {
  __typename?: 'User';
  email: Scalars['String']['output'];
  name?: Maybe<Scalars['String']['output']>;
};

export type RemoteFileSyncEventsSubscriptionVariables = Exact<{
  localFilesVersion: Scalars['String']['input'];
}>;


export type RemoteFileSyncEventsSubscription = { __typename?: 'Subscription', remoteFileSyncEvents: { __typename?: 'RemoteFileSyncEvents', remoteFilesVersion: string, changed: Array<{ __typename?: 'FileSyncChangedEvent', path: string, mode: number, content: string, encoding: FileSyncEncoding }>, deleted: Array<{ __typename?: 'FileSyncDeletedEvent', path: string }> } };

export type RemoteFilesVersionQueryVariables = Exact<{ [key: string]: never; }>;


export type RemoteFilesVersionQuery = { __typename?: 'Query', remoteFilesVersion: string };

export type PublishFileSyncEventsMutationVariables = Exact<{
  input: PublishFileSyncEventsInput;
}>;


export type PublishFileSyncEventsMutation = { __typename?: 'Mutation', publishFileSyncEvents: { __typename?: 'PublishFileSyncEventsResult', remoteFilesVersion: string } };
