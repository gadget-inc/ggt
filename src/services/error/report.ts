import * as Sentry from "@sentry/node";
import ms from "ms";
import os from "node:os";
import { serializeError } from "serialize-error";
import type { App } from "../app/app.js";
import { config } from "../config/config.js";
import { env } from "../config/env.js";
import { createLogger } from "../output/log/logger.js";
import type { User } from "../user/user.js";
import { parseBoolean } from "../util/boolean.js";
import { CLIError, IsBug } from "./error.js";

const log = createLogger({ name: "report" });

let user: User | undefined;
export const setUser = (newUser: typeof user): void => {
  log.trace("set user", { user: newUser });
  user = newUser;
  // eslint-disable-next-line unicorn/no-null
  Sentry.setUser(newUser ?? null);
};

let app: App | undefined;
export const setApp = (newApp: typeof app): void => {
  log.trace("set app", { app: newApp });
  app = newApp;
};

export const reportErrorAndExit = async (cause: unknown): Promise<never> => {
  log.error("reporting error and exiting", { error: cause });

  try {
    const error = CLIError.from(cause);
    log.println(error.render());

    if (error.isBug !== IsBug.NO) {
      Sentry.getCurrentHub().captureException(error, {
        event_id: error.sentryEventId,
        captureContext: {
          user: user ? { id: String(user.id), email: user.email, username: user.name ?? undefined } : undefined,
          tags: {
            applicationId: app ? app.id : undefined,
            arch: config.arch,
            bug: error.isBug,
            code: error.code,
            environment: env.value,
            platform: config.platform,
            shell: config.shell,
            version: config.version,
          },
          contexts: {
            cause: error.cause ? serializeError(error.cause) : undefined,
            app: {
              command: `ggt ${process.argv.slice(2).join(" ")}`,
              argv: process.argv,
            },
            device: {
              name: os.hostname(),
              family: os.type(),
              arch: os.arch(),
            },
            runtime: {
              name: process.release.name,
              version: process.version,
            },
          },
        },
      });

      await Sentry.flush(ms("2s"));
    }
  } finally {
    process.exit(1);
  }
};

export const installErrorHandlers = (): void => {
  log.debug("installing error handlers");

  Sentry.init({
    dsn: "https://0c26e0d8afd94e77a88ee1c3aa9e7065@o250689.ingest.sentry.io/6703266",
    release: config.version,
    enabled: env.productionLike && parseBoolean(process.env["GGT_SENTRY_ENABLED"] ?? "true"),
  });

  process.once("uncaughtException", (error) => {
    void reportErrorAndExit(error);
  });

  process.once("unhandledRejection", (error) => {
    void reportErrorAndExit(error);
  });
};
