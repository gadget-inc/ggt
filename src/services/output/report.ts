import * as Sentry from "@sentry/node";
import cleanStack from "clean-stack";
import ms from "ms";
import { randomUUID } from "node:crypto";
import os from "node:os";
import type { Context } from "../command/context.js";
import { config } from "../config/config.js";
import { env } from "../config/env.js";
import { parseBoolean } from "../util/boolean.js";
import { serializeError } from "../util/object.js";
import { workspaceRoot } from "../util/paths.js";
import { sprint } from "./sprint.js";

export const reportErrorAndExit = async (ctx: Context, cause: unknown): Promise<never> => {
  ctx.log.error("reporting error and exiting", { error: cause });

  try {
    const error = CLIError.from(cause);
    ctx.log.println({ marginTop: true })(error.toString());

    if (error.isBug === IsBug.NO) {
      return undefined as never;
    }

    Sentry.captureException(error, {
      event_id: error.id,
      captureContext: {
        user: ctx.user && {
          id: String(ctx.user.id),
          email: ctx.user.email,
          username: ctx.user.name ?? undefined,
        },
        tags: {
          application_id: ctx.app?.id,
          arch: config.arch,
          bug: error.isBug,
          environment: env.value,
          platform: config.platform,
          shell: config.shell,
          version: config.version,
        },
        contexts: {
          ctx: {
            argv: process.argv,
            args: ctx.args,
          },
          cause: error.cause ? serializeError(error.cause) : undefined,
          app: {
            app_name: config.name,
            app_version: config.version,
          },
          device: {
            name: os.hostname(),
            family: os.type(),
            arch: os.arch(),
          },
          runtime: {
            name: process.release.name,
            version: process.version,
          },
        },
      },
    });

    await Sentry.flush(ms("2s"));
  } finally {
    process.exit(1);
  }
};

export const installErrorHandlers = (ctx: Context): void => {
  ctx.log.debug("installing error handlers");

  Sentry.init({
    dsn: "https://0c26e0d8afd94e77a88ee1c3aa9e7065@o250689.ingest.sentry.io/6703266",
    release: config.version,
    enabled: env.productionLike && parseBoolean(process.env["GGT_SENTRY_ENABLED"] ?? "true"),
    environment: config.version.includes("experimental") ? "experimental" : "production",
  });

  const handleError = (error: unknown) => void reportErrorAndExit(ctx, error);
  process.once("uncaughtException", handleError);
  process.once("unhandledRejection", handleError);
};

export const IsBug = Object.freeze({
  YES: "yes",
  NO: "no",
  MAYBE: "maybe",
});

export type IsBug = (typeof IsBug)[keyof typeof IsBug];

/**
 * Base class for all errors.
 */
export abstract class CLIError extends Error {
  /**
   * The ID for this error.
   */
  id = env.testLike ? "00000000-0000-0000-0000-000000000000" : randomUUID();

  /**
   * The underlying *thing* that caused this error.
   */
  cause?: unknown;

  /**
   * Assume the stack trace exists.
   */
  override stack!: string;

  /**
   * Indicates whether this error is considered a bug or not.
   */
  abstract isBug: IsBug;

  constructor(message: string) {
    super(message);
    Error.captureStackTrace(this, this.constructor);
    this.stack = cleanStack(this.stack, { pretty: true, basePath: workspaceRoot });
  }

  /**
   * Constructs a CLIError from an unknown cause.
   *
   * @param cause - The cause of the error.
   */
  static from(cause: unknown): CLIError {
    if (cause instanceof CLIError) {
      return cause;
    }
    return new UnexpectedError(cause);
  }

  override toString(): string {
    let rendered = this.render();

    if (this.isBug !== IsBug.NO) {
      rendered +=
        "\n\n" +
        sprint`
          ${this.isBug === IsBug.YES ? "This is a bug" : "If you think this is a bug"}, please submit an issue using the link below.

          https://github.com/gadget-inc/ggt/issues/new?template=bug_report.yml&error-id=${this.id}
        `;
    }

    return rendered;
  }

  /**
   * Turns this error into a user-friendly message that explains what
   * went wrong and how to fix it. A good write up of what an error
   * should look like can be found here:
   * {@link https://clig.dev/#errors}
   */
  protected abstract render(): string;
}

/**
 * Our "catch all" error.
 *
 * If this error is thrown, we almost certainly have a bug, and should
 * either fix it or add a more specific error so that we can provide
 * more useful information.
 */
export class UnexpectedError extends CLIError {
  isBug = IsBug.YES;

  constructor(override cause: unknown) {
    super("An unexpected error occurred");
  }

  protected render(): string {
    const serialized = serializeError(this.cause);
    const body = serialized.stack || serialized.message || this.stack;
    return this.message + "\n\n" + body;
  }
}
