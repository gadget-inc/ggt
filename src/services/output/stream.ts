import process from "node:process";
import { env } from "../config/env.js";
import { isObject } from "../util/is.js";

/**
 * A wrapper around process.stdout and process.stderr that allows us to mock out the streams for testing.
 *
 * @see https://github.com/oclif/core/blob/16139fe8a7f991b4b446a1599ab63f15d9809b8e/src/cli-ux/stream.ts
 */
export class Stream {
  /**
   * Indicates whether the last line that was written was empty. This is
   * used to prevent writing more than one empty line in a row.
   */
  private _lastLineWasEmpty = true;

  constructor(public channel: "stdout" | "stderr") {
    process[this.channel].on("error", (err: unknown) => {
      if (isObject(err) && "code" in err && err.code === "EPIPE") {
        return;
      }
      throw err;
    });
  }

  get isTTY(): boolean {
    return process[this.channel].isTTY;
  }

  getWindowSize(): number[] {
    return process[this.channel].getWindowSize();
  }

  write(str: string): void {
    // remove duplicate empty lines
    while (str.startsWith("\n\n")) {
      str = str.slice(0, -1);
    }

    while (str.endsWith("\n\n")) {
      str = str.slice(0, -1);
    }

    str = str.replaceAll(/\n\n+/g, "\n\n");

    if (this._lastLineWasEmpty) {
      // we just printed an empty line, so don't print another one
      while (str.startsWith("\n")) {
        str = str.slice(1);
      }
    }

    if (str === "") {
      // nothing to print
      return;
    }

    // remember if the last line was empty
    this._lastLineWasEmpty = str === "\n";

    this._write(str);
  }

  on(event: string, listener: (...args: unknown[]) => void): this {
    process[this.channel].on(event, listener);
    return this;
  }

  once(event: string, listener: (...args: unknown[]) => void): this {
    process[this.channel].once(event, listener);
    return this;
  }

  private _write(data: string): boolean {
    if (env.testLike) {
      // use console.log/error in tests since vitest doesn't display
      // process.stdout/stderr correctly; also, remove trailing newline
      // since console.log/error adds one.
      data = data.replace(/\n$/, "");
      if (this.channel === "stdout") {
        console.log(data);
      } else {
        console.error(data);
      }
      return true;
    }

    return process[this.channel].write(data);
  }
}

export const stdout = new Stream("stdout");
export const stderr = new Stream("stderr");
