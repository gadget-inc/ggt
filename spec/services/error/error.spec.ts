import { GraphQLError } from "graphql";
import { dedent } from "ts-dedent";
import { describe, expect, it } from "vitest";
import type { CloseEvent, ErrorEvent } from "ws";
import {
  CLIError,
  EditGraphQLError,
  InvalidSyncFileError,
  IsBug,
  UnexpectedError,
  YarnNotFoundError,
} from "../../../src/services/error/error.js";
import { reportErrorAndExit } from "../../../src/services/error/report.js";
import { expectProcessExit } from "../../__support__/process.js";
import { expectStdout } from "../../__support__/stream.js";

describe("reportErrorAndExit", () => {
  it("captures errors", async () => {
    class TestError extends CLIError {
      override isBug = IsBug.NO;

      constructor() {
        super("GGT_CLI_TEST_ERROR", "Boom!");
      }

      protected override body(): string {
        return this.message;
      }
    }

    const error = new TestError();

    await expectProcessExit(() => reportErrorAndExit(error), 1);

    expectStdout().toEqual(
      dedent`
        GGT_CLI_TEST_ERROR: Boom!

        Boom!\n
      `,
    );
  });
});

describe("UnexpectedError", () => {
  it("renders correctly", () => {
    const cause = new Error("Whoops!");
    cause.stack = "Error: Whoops!\n    at <anonymous>:1:1";
    const error = new UnexpectedError(cause);
    expect(error.render()).toMatchSnapshot();
  });
});

describe("EditGraphQLError", () => {
  it("renders a GraphQL error correctly", () => {
    const error = new EditGraphQLError("query { foo }", [new GraphQLError("Changed and deleted files must not overlap")]);
    expect(error.render()).toMatchSnapshot();
  });

  it("renders multiple GraphQL errors correctly", () => {
    const error = new EditGraphQLError("query { foo }", [
      new GraphQLError("Changed and deleted files must not overlap"),
      new GraphQLError("Files version mismatch, expected 1 but got 2"),
    ]);
    expect(error.render()).toMatchSnapshot();
  });

  it("renders a CloseEvent correctly", () => {
    const error = new EditGraphQLError("query { foo }", {
      type: "close",
      code: 1000,
      reason: "Normal closure",
      wasClean: true,
    } as CloseEvent);
    expect(error.render()).toMatchSnapshot();
  });

  it("renders an ErrorEvent correctly", () => {
    const error = new EditGraphQLError("query { foo }", {
      type: "error",
      message: "connect ECONNREFUSED 10.254.254.254:3000",
      error: {
        errno: -61,
        code: "ECONNREFUSED",
        syscall: "connect",
        address: "10.254.254.254",
        port: 3000,
      },
    } as ErrorEvent);
    expect(error.render()).toMatchSnapshot();
  });

  it("renders a string correctly", () => {
    const error = new EditGraphQLError("query { foo }", "We received a response without data");
    expect(error.render()).toMatchSnapshot();
  });
});

describe("YarnNotFoundError", () => {
  it("renders correctly", () => {
    const error = new YarnNotFoundError();
    expect(error.render()).toMatchSnapshot();
  });
});

describe("InvalidSyncFileError", () => {
  it("renders correctly", () => {
    const dir = "~/gadget/test";
    const app = "test";

    const error = new InvalidSyncFileError(dir, app);
    expect(error.render()).toMatchSnapshot();
  });
});
