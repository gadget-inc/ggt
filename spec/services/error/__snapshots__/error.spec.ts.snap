// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditGraphQLError > renders a CloseEvent correctly 1`] = `
"GGT_CLI_CLIENT_ERROR: An error occurred while communicating with Gadget

The connection to Gadget closed unexpectedly.

If you think this is a bug, please submit an issue using the link below.

https://github.com/gadget-inc/ggt/issues/new?template=bug_report.yml&error-id=00000000-0000-0000-0000-000000000000"
`;

exports[`EditGraphQLError > renders a GraphQL error correctly 1`] = `
"GGT_CLI_CLIENT_ERROR: An error occurred while communicating with Gadget

Gadget responded with the following error:

  Changed and deleted files must not overlap

If you think this is a bug, please submit an issue using the link below.

https://github.com/gadget-inc/ggt/issues/new?template=bug_report.yml&error-id=00000000-0000-0000-0000-000000000000"
`;

exports[`EditGraphQLError > renders a string correctly 1`] = `
"GGT_CLI_CLIENT_ERROR: An error occurred while communicating with Gadget

We received a response without data

If you think this is a bug, please submit an issue using the link below.

https://github.com/gadget-inc/ggt/issues/new?template=bug_report.yml&error-id=00000000-0000-0000-0000-000000000000"
`;

exports[`EditGraphQLError > renders an ErrorEvent correctly 1`] = `
"GGT_CLI_CLIENT_ERROR: An error occurred while communicating with Gadget

connect ECONNREFUSED 10.254.254.254:3000

If you think this is a bug, please submit an issue using the link below.

https://github.com/gadget-inc/ggt/issues/new?template=bug_report.yml&error-id=00000000-0000-0000-0000-000000000000"
`;

exports[`EditGraphQLError > renders multiple GraphQL errors correctly 1`] = `
"GGT_CLI_CLIENT_ERROR: An error occurred while communicating with Gadget

Gadget responded with multiple errors:

  1. Changed and deleted files must not overlap
  2. Files version mismatch, expected 1 but got 2

If you think this is a bug, please submit an issue using the link below.

https://github.com/gadget-inc/ggt/issues/new?template=bug_report.yml&error-id=00000000-0000-0000-0000-000000000000"
`;

exports[`InvalidSyncFileError > renders correctly 1`] = `
"GGT_CLI_INVALID_SYNC_FILE: The .gadget/sync.json file was invalid or not found

We failed to read the Gadget metadata file in this directory:

  ~/gadget/test

If you're running \`ggt sync\` for the first time, we recommend using an empty directory such as:

  ~/gadget/test

Otherwise, if you're sure you want to sync the contents of that directory to Gadget, run \`ggt sync\` again with the \`--force\` flag:

  $ ggt sync --app test ~/gadget/test --force

You will be prompted to either merge your local files with your remote ones or reset your local files to your remote ones.

If you think this is a bug, please submit an issue using the link below.

https://github.com/gadget-inc/ggt/issues/new?template=bug_report.yml&error-id=00000000-0000-0000-0000-000000000000"
`;

exports[`UnexpectedError > renders correctly 1`] = `
"GGT_CLI_UNEXPECTED_ERROR: An unexpected error occurred

Error: Whoops!
    at <anonymous>:1:1

This is a bug, please submit an issue using the link below.

https://github.com/gadget-inc/ggt/issues/new?template=bug_report.yml&error-id=00000000-0000-0000-0000-000000000000"
`;

exports[`YarnNotFoundError > renders correctly 1`] = `
"GGT_CLI_YARN_NOT_FOUND: Yarn not found

Yarn must be installed to sync your application. You can install it by running:

  $ npm install --global yarn

For more information, see: https://classic.yarnpkg.com/en/docs/install"
`;
